layer/migration.go

```diff
@@ -77,79 +76,72 @@ func (ls *layerStore) MountByGraphID(name string, graphID string, parent ChainID
 	return m, nil
 }
 
-func (ls *layerStore) migrateLayer(tx MetadataTransaction, tarDataFile string, layer *roLayer) error {
-	var ar io.Reader
-	var tdf *os.File
-	var err error
-	if tarDataFile != "" {
-		tdf, err = os.Open(tarDataFile)
-		if err != nil {
-			if !os.IsNotExist(err) {
-				return err
-			}
-			tdf = nil
-		}
-		defer tdf.Close()
-	}
-	if tdf != nil {
-		tsw, err := tx.TarSplitWriter()
+func (ls *layerStore) ChecksumForGraphID(id, parent, oldTarDataPath, newTarDataPath string) (diffID DiffID, size int64, err error) {
+	defer func() {
 		if err != nil {
-			return err
+			logrus.Debugf("could not get checksum for %q with tar-split: %q", id, err)
+			diffID, size, err = ls.checksumForGraphIDNoTarsplit(id, parent, newTarDataPath)
 		}
+	}()
 
-		defer tsw.Close()
-
-		uncompressed, err := gzip.NewReader(tdf)
-		if err != nil {
-			return err
-		}
-		defer uncompressed.Close()
+	if oldTarDataPath == "" {
+		err = errors.New("no tar-split file")
+		return
+	}
 
-		tr := io.TeeReader(uncompressed, tsw)
-		trc := ioutils.NewReadCloserWrapper(tr, uncompressed.Close)
+	tarDataFile, err := os.Open(oldTarDataPath)
+	if err != nil {
+		return
+	}
+	defer tarDataFile.Close()
+	uncompressed, err := gzip.NewReader(tarDataFile)
+	if err != nil {
+		return
+	}
 
-		ar, err = ls.assembleTar(layer.cacheID, trc, &layer.size)
-		if err != nil {
-			return err
-		}
+	dgst := digest.Canonical.New()
+	err = ls.assembleTarTo(id, uncompressed, &size, dgst.Hash())
+	if err != nil {
+		return
+	}
 
-	} else {
-		var graphParent string
-		if layer.parent != nil {
-			graphParent = layer.parent.cacheID
-		}
-		archiver, err := ls.driver.Diff(layer.cacheID, graphParent)
-		if err != nil {
-			return err
-		}
-		defer archiver.Close()
+	diffID = DiffID(dgst.Digest())
+	os.RemoveAll(newTarDataPath)
```

Comment:
Is the expected failure mode that the file doesn't exist? Maybe log unless `os.IsNotExist(err)`.